onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -503,6 +536,48 @@ void UpdateCudaProviderOptions(InferenceSession* sess, onnxruntime::CudaProvider
     LOGS(*(sess->GetLogger()), INFO) << "cuda arean extend strategy is set to " << it->second;
   }
 }
+
+AllocatorPtr GetCudaAllocator(OrtDevice::DeviceId id) {
+  // Current approach is not thread-safe, but there are some bigger infra pieces to put together in order to make
+  // multi-threaded CUDA allocation work we need to maintain a per-thread CUDA allocator
+  static std::unordered_map<OrtDevice::DeviceId, AllocatorPtr> id_to_allocator_map;
+
+  if (id_to_allocator_map.find(id) == id_to_allocator_map.end()) {
+    // Use arena-based allocator
+    AllocatorCreationInfo default_memory_info(
+        [](OrtDevice::DeviceId id) {
+          return onnxruntime::make_unique<CUDAAllocator>(id, CUDA);
+        },
+        id,
+        true,
+        {cuda_mem_limit,
+         static_cast<int>(arena_extend_strategy),
+         -1, -1});
+
+    auto allocator = CreateAllocator(default_memory_info);
+
+    id_to_allocator_map.insert({id, allocator});
+  }
```

Comment:
Do we want to have another allocator with another arena just to support usage when there's no inference session (so that you can get an existing allocator from the session state)? 

If we support that usage, will we need to create allocators for more device types?

Reply:
We need some mechanism to allocate memory on a device in case the user requests that the memory backing an OrtValue reside on the device for which we need to get hold of an allocator. Since OrtValues are conceptually de-coupled for a session, we have no choice but to create an allocator for use here. Do you have any other approach in mind ? 

Yes, we will need to create allocators corresponding to each device type that we support for OrtValues.